.gitlab-ci.yml

```diff
@@ -91,6 +91,8 @@ test-linux-stable:                 &test
   except:
     variables:
       - $DEPLOY_TAG
+  before_script:
+    - git fetch origin && git merge origin/master
```

Comment:
I don't think this will work like that. If there is a merge conflict your test will fail. One would need something like https://github.com/paritytech/substrate/blob/v1.0.0rc1/scripts/gitlab/check_merge_conflict.sh#L75

I can imagine that if you're developing on a feature you probably want to focus on your task and not tackle with the master merge conflict after every commit.

Reply:
Yeah, that is the whole point of this feature. Your pr should fail if merging is not possible. If you develop a feature, you should test it on your own. Waiting everytime for the CI to build your stuff sounds like you wait most of the time.

Reply:
We had yesterday a failing master, which would have probably been catched by merging the master before running tests. Sometimes there is no merge conflict, but a type change that prevents your pr from compiling on master.

Projects like rust are doing the same to prevent these kind of errors.

Reply:
But if you do not immediately merge after that it will lead to the wrong impression that it has been tested on the master branch - which may not be the case if other pull requests have been merged in the meantime (this happend several times with the wasm blobs).

Are there situations where the tests fail on master after they have been run successfully on the pull request?

Reply:
forget my last question.

Reply:
What about allowing only fast-forward merges as mentioned by @tomaka so devs would need to merge in master themselves before they want to merge.
(Otherwise the workflow will significantly change: If you push a commit and want to add another one you'll always have to fetch your branch first from github - or you accept having another merge for that).

Reply:
It's true that it would be quite annoying to have to merge master in order for CI to run.
To me the workflow is that we should have CI run normally while the PR gets reviewed, then once we decide the PR is good, we check whether it still runs on top of master.

Reply:
How do we check that the PR still runs on top of master? Manually?

> (Otherwise the workflow will significantly change: If you push a commit and want to add another one you'll always have to fetch your branch first from github - or you accept having another merge for that).

That is not true, as we not push the merge back into the PR. We just merge in CI and do not push back to github.

Reply:
Why not enable the branch protection feature of `Require branches to be up-to-date before merging`?

Reply:
I think if we require branches to be up-to-date with master, we will make it more worse than with the WASM blobs. 
With WASM blobs we already had the problem that people did not update their stuff and we had a CI job to automatically rebuild merge, rebuild and push.

Having merge conflicts with master is not that often I would say and if so, merging master should not be that hard. I don't see that it produces that much hassle for people.

Reply:
Merge conflicts with master is almost always the case: after a PR is merged that touches anything to do with the runtime (which is 80% of PRs), then the runtime version gets bumped. This instantly causes a conflict with all other such PRs.

Reply:
sure, but I want CI to run first. bumping the version is the very last thing to do.

Reply:
Okay.

Then we probably need a bot to solve this while merging.